data/domain/tome.tsx

```diff
@@ -0,0 +1,896 @@
+                if (this.currentValue > 5 * this.data.keyQty) {
+                    return 0;
+                } else {
+                    return Math.pow(1.2 * (6 * this.data.keyQty - this.currentValue) / (7 * this.data.keyQty - this.currentValue), 5);
+                }
+            default:
+                return 0;
+        }
+    }
+
+    getLineDescription = (): string => {
+        return this.data.desc?.replace('(Tap for more info)', '').replace('膛', '').trim() ?? '';
+    }
+
+    getLineName = (): string => {
+        return this.data.name?.replace('(Tap for more info)', '').replace('膛', '').trim() ?? '';
+    }
+
+    private getLineScore = (): number => {
+        return Math.ceil(this.getLineMultiplyer() * this.data.totalVal);
+    }
+
+    getCurrentValueDisplay = (): string => {
+        switch(this.index) {
+            // Big values
+            case 8:
+            case 9:
+            case 13:
+            case 14:
+            case 31:
+            case 32:
+            case 33:
+            case 34:
+            case 35:
+            case 46:
+            case 53:
+            case 61:
+            case 64:
+            case 66:
+            case 78:
+                return nFormatter(this.currentValue);
+            // Not so big values but with lots of decimals and wanna keep a bit of it
+            case 16:
+            case 18:
+                return (Math.round(100 * this.currentValue) / 100).toString();
+            default:
+                return this.currentValue.toString();
+        }
+    }
+
+    getLineDisplayColor = (): string => {
+        switch(true) {
+            case this.lineScore >= this.data.totalVal:
+                return TomeScoreColors.Platinum;
+            case this.lineScore >= this.data.totalVal*0.75:
+                return TomeScoreColors.Gold;
+            case this.lineScore >= this.data.totalVal*0.5:
+                return TomeScoreColors.Silver;
+            default:
+                return TomeScoreColors.Bronze;
+        }
+    }
+
+    getLineUnlockLevel = (): number => {
+        return (500 + (50 * TomeLineUnlockLevel[this.index] + (10 * Math.max(0, TomeLineUnlockLevel[this.index] - 30) + 10 * Math.max(0, TomeLineUnlockLevel[this.index] - 50))));
+    }
+
+    private updateIsLineUnlocked = () => {
+        this.unlocked = this.accountTotalLevel >= this.getLineUnlockLevel();
+    }
+
+    updateCurrentValue = (value: number, totalAccountLevel: number) => {
+        if (this.currentValue != value) {
+            this.currentValue = value;
+        }
+        if (this.accountTotalLevel != totalAccountLevel) {
+            this.accountTotalLevel = totalAccountLevel;
+            this.updateIsLineUnlocked();
+        }
+        this.updateLineScore();
+    }
+
+    private updateLineScore = () => {
+        this.lineScore = this.unlocked ? this.getLineScore() : 0;
+    }
+}
+
+export class Tome extends Domain {
+    // Per player
+    lines: TomeLine[][] = [];
```

Comment:
Will the function / logic / calculations work if this was an array of lines and inside each line you had each player's score?

I find this approach somewhat not ideal and wondering if we can change it.

My reasoning is as such:
* You are duplicating all the line data per player, instead of just keeping a numeric value of the player score inside the line
* You are tracking if a line is unlocked or not across every player
* You are tracking the total account level per line per player, instead of the for the entire tome which is counter intuitive in my mind

Now there might be something I'm missing with the way the calculations work and how figuring out the highest works but I would like to think we can do this change but still get all the relevant math done correctly.

What do you think?

Reply:
I think you are indeed right

I was pretty head down on this and maybe not though about it enough before deciding to handle this in the current way, it shouldn't be that bad to change it, so gonna try to change it after addressing all other comments

Reply:
Now that I'm looking at it, I'll also need to store the "current value" of each line per player, as it's this variable that changes from player to player, making the score vary

Reply:
I made some changes, currentValues and lineScores are now arrays with a value for each player

Let me know what you think about those changes, hopefully it's better now 😄